message-broker/src/lib.rs

```diff
@@ -86,6 +141,50 @@ decl_module! {
 				);
 				*up = up.split_off(num);
 			});
+
+			// Sending HRMP messages is a little bit more involved. On top of the number of messages
+			// per block limit, there is also a constraint that it's possible to send only a single
+			// message to a given recipient per candidate.
+			let mut non_empty_hrmp_channels = <Self as Store>::NonEmptyHrmpChannels::get();
+			let outbound_hrmp_num = cmp::min(HRMP_MSG_NUM_PER_CANDIDATE, non_empty_hrmp_channels.len());
+			let mut outbound_hrmp_messages = Vec::with_capacity(outbound_hrmp_num);
+			let mut prune_empty = Vec::with_capacity(outbound_hrmp_num);
+
+			for &recipient in non_empty_hrmp_channels.iter().take(outbound_hrmp_num) {
+				let (message_payload, became_empty) =
+					<Self as Store>::OutboundHrmpMessages::mutate(&recipient, |v| {
+						// this panics if `v` is empty. However, we are iterating only once over non-empty
+						// channels, therefore it cannot panic.
+						let first = v.remove(0);
+						let became_empty = v.is_empty();
+						(first, became_empty)
+					});
+
+				outbound_hrmp_messages.push(OutboundHrmpMessage {
+					recipient,
+					data: message_payload,
+				});
+				if became_empty {
+					prune_empty.push(recipient);
+				}
+			}
+
+			// Prune hrmp channels that became empty. Additionally, because it may so happen that we
+			// only gave attention to some channels in `non_empty_hrmp_channels` it's important to
+			// change the order. Otherwise, the next `on_finalize` we will again give attention
+			// only to those channels that happen to be in the beginning, until they are emptied.
+			// This leads to "starvation" of the channels near to the end.
+			//
+			// To mitigate this we shift all processed elements towards the end of the vector using
+			// `rotate_left`. To get intution how it works see the examples in its rustdoc.
+			non_empty_hrmp_channels.retain(|x| !prune_empty.contains(x));
+			non_empty_hrmp_channels.rotate_left(outbound_hrmp_num - prune_empty.len());
```

Comment:
Doesn't this breaks the `binary_search` we do when inserting into this vector?

Reply:
this does, good point! We don't need it to be sorted though.